openlibrary/plugins/upstream/code.py

```diff
@@ -215,7 +215,8 @@ def POST(self, key):
             raise web.seeother(web.changequery({}))
 
         user = web.ctx.site.get_user()
-        if not user and (user.is_admin() or user.is_librarian() and web.ctx.site.can_write(key)):
+        if (not user or not web.ctx.site.can_write(key) or
+                not (user.is_admin() or user.is_librarian())):
```

Comment:
This is a difficult construct to reason about which could lead to further errors.

Please create a separate function like:
```python
def user_is_admin_or_librarian():
    user = web.ctx.site.get_user()
    return user and (user.is_admin() or user.is_librarian())
```
This will make it much easier to read and write the remaining if statement.

Reply:
Done.

Reply:
And I re-confirmed that a logged out user and a logged-in, non-admin, non-librarian user get a permission denied page.  I also verified that an admin can still revert:

![image](https://user-images.githubusercontent.com/72705998/99354388-34ecc800-285b-11eb-95c2-f7068bae0814.png)